File: src/test/java/br/com/blz/testjava/business/ProductBusinessTest.java
package br.com.blz.testjava.business;

import br.com.blz.testjava.TestJavaApplication;
import br.com.blz.testjava.domain.Product;
import br.com.blz.testjava.enums.WarehouseType;
import br.com.blz.testjava.exception.SkuExistsException;
import br.com.blz.testjava.exception.SkuNotExistsException;
import br.com.blz.testjava.repository.ProductRepository;
import br.com.blz.testjava.service.ProductService;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import java.util.List;

import static br.com.blz.testjava.business.util.MockObjects.*;
import static org.junit.jupiter.api.Assertions.*;

@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = {ProductBusiness.class, ProductService.class, ProductRepository.class, TestJavaApplication.class})
class ProductBusinessTest {

    @Autowired
    private ProductBusiness productBusiness;

    @Autowired
    private ProductService productService;

    @Autowired
    private ProductRepository productRepository;

    @BeforeEach
    public void before() {
        this.productRepository.deleteAll();
    }

    @Test
    void mustSavedProductSuccessfullyAndReturnsMarketableTrueAndInvetoryQuantity() {
        var warehouses = List.of(createWarehouse("SP", WarehouseType.PHYSICAL_STORE,
            250l));
        var inventory = createInventory(warehouses);
        var product = createProduct("ps5 digital edition", 12234l, inventory);

        var productReturn = this.productBusiness.save(product);

        assertAll(
            () -> assertTrue(productReturn.isMarketable()),
            () -> assertEquals(250l, productReturn.getInventory().getQuantity()));
    }

    @Test
    void mustSavedProductSuccessfullyAndReturnisMarketableFalseAndInvetoryQuantity() {
        var inventory = createInventory(null);
        var product = createProduct("ps5 digital edition", 12234l, inventory);

        var productReturn = this.productBusiness.save(product);

        assertAll(
            () -> assertFalse(productReturn.isMarketable()),
            () -> assertEquals(0, productReturn.getInventory().getQuantity()));
    }


    @Test
    void mustThrowExceptionBecauseProductExistsInBase() {
        var warehouses = List.of(createWarehouse("SP", WarehouseType.PHYSICAL_STORE,
            250l));
        var inventory = createInventory(warehouses);
        var product = createProduct("ps5 digital edition", 12234l, inventory);

        this.productBusiness.save(product);

        assertThrows(SkuExistsException.class, () -> this.productBusiness.save(product));

    }

    @Test
    void mustReturnsProductBySku() {
        var warehouses = List.of(createWarehouse("SP", WarehouseType.PHYSICAL_STORE,
            250l));
        var inventory = createInventory(warehouses);
        var product = createProduct("ps5 digital edition", 12234l, inventory);

        this.productBusiness.save(product);
        var productReturn = this.productBusiness.findBySku(12234l);

        Assertions.assertAll(
            () -> Assertions.assertTrue(productReturn.isMarketable()),
            () -> Assertions.assertEquals(250l, productReturn.getInventory().getQuantity()),
            () -> Assertions.assertEquals(product.getName(), productReturn.getName()));
    }

    @Test
    void mustThrowExceptionBecauseProductBySkuNotExistsInBase() {
        Assertions.assertThrows(SkuNotExistsException.class, () -> this.productBusiness.findBySku(12234l));
    }

    @Test
    void deveEstourarExcecaoAoFazerUpdatePoisProdutoNaoExiste() {
        Assertions.assertThrows(SkuNotExistsException.class, () -> this.productBusiness.update(12234l,
            Product.builder().build()));
    }

    @Test
    void deveAlterarProdutoComSucesso() {
        var warehouses = List.of(createWarehouse("SP", WarehouseType.PHYSICAL_STORE,
            250l));
        var inventory = createInventory(warehouses);
        var product = createProduct("ps5 digital edition", 12234l, inventory);

        var newWarehouses = List.of(createWarehouse("SP", WarehouseType.PHYSICAL_STORE,
            250l), createWarehouse("SC", WarehouseType.PHYSICAL_STORE,
            250l));
        var newInventory = createInventory(newWarehouses);
        var newProduct = createProduct("ps5 digital edition", 12234l, newInventory);

        this.productBusiness.save(product);
        var newProductSaved = this.productBusiness.update(12234l, newProduct);

        Assertions.assertAll(
            () -> Assertions.assertTrue(newProductSaved.isMarketable()),
            () -> Assertions.assertEquals(500l, newProductSaved.getInventory().getQuantity()),
            () -> Assertions.assertEquals(2, newProductSaved.getInventory().getWarehouses().size()),
            () -> Assertions.assertEquals(newProduct.getName(), newProductSaved.getName()));
    }

}

File: src/test/java/br/com/blz/testjava/product/business/ProductBusinessTest.kt
package br.com.blz.testjava.product.business

import br.com.blz.testjava.product.ProductTestTemplates
import br.com.blz.testjava.product.business.objects.Product
import br.com.blz.testjava.product.business.objects.ProductInventory
import br.com.blz.testjava.product.exception.ProductNotFoundException
import br.com.blz.testjava.product.exception.ProductSkuDuplicatedException
import br.com.blz.testjava.product.repository.ProductRepository
import org.junit.Test
import org.junit.jupiter.api.assertThrows
import org.junit.runner.RunWith
import org.springframework.test.context.junit4.SpringRunner
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

@RunWith(SpringRunner::class)
class ProductBusinessTest {

  private val productBusiness = ProductBusiness()
  @Test
  fun `Avoid save duplicaded sku`() {
    val product = ProductTestTemplates.createProduct()

    productBusiness.create(product)
    assertThrows<ProductSkuDuplicatedException> { productBusiness.create(product) }
  }

  @Test
  fun `Throw exception updating not existent product`() {
    val product = ProductTestTemplates.createProduct(2L, "Product 2")

    assertThrows<ProductNotFoundException> { ProductBusiness().update(product) }
  }

  @Test
  fun `Update product name`() {
    ProductBusiness().create(ProductTestTemplates.createProduct(3L, "Product 3"))
    ProductBusiness().update(ProductTestTemplates.createProduct(3L, "Product 03"))

    val changedProduct = ProductRepository.get(3L)

    assertEquals("Product 03", changedProduct!!.name)
  }

  @Test
  fun `Remove product`() {
    ProductBusiness().create(ProductTestTemplates.createProduct(4L, "Product 3"))
    assertNotNull(ProductRepository.get(4L))

    ProductBusiness().delete(4L)
    assertNull(ProductRepository.get(4L))
  }
}
